hbase-client/src/main/java/org/apache/hadoop/hbase/client/CheckAndMutate.java

```diff
@@ -312,83 +331,178 @@ public Row getAction() {
-    }
-    throw new UnsupportedOperationException();
+    throw new UnsupportedOperationException("Please call this method of the individual mutations");
   }
 
   @Override
   public long getTimestamp() {
-    if (action instanceof Mutation) {
-      return ((Mutation) action).getTimestamp();
-    }
-    throw new UnsupportedOperationException();
+    throw new UnsupportedOperationException("Please call this method of the individual mutations");
   }
 
   @Override
-  public Mutation setTimestamp(long timestamp) {
-    if (action instanceof Mutation) {
-      return ((Mutation) action).setTimestamp(timestamp);
-    }
-    throw new UnsupportedOperationException();
+  public CheckAndMutate setTimestamp(long timestamp) {
+    throw new UnsupportedOperationException("Please call this method of the individual mutations");
   }
 
+  /**
+   * @return the highest durability of all the mutations
+   */
   @Override
   public Durability getDurability() {
-    if (action instanceof Mutation) {
-      return ((Mutation) action).getDurability();
+    if (initDurability) {
+      return super.getDurability();
     }
-    throw new UnsupportedOperationException();
+    initDurability = true;
+    for (Mutation mutation : getMutations()) {
+      Durability tmpDur = mutation.getDurability();
+      if (tmpDur.ordinal() > super.getDurability().ordinal()) {
+        super.setDurability(tmpDur);
+      }
+    }
+    return super.getDurability();
   }
 
   @Override
-  public Mutation setDurability(Durability d) {
-    if (action instanceof Mutation) {
-      return ((Mutation) action).setDurability(d);
-    }
-    throw new UnsupportedOperationException();
+  public CheckAndMutate setDurability(Durability d) {
+    throw new UnsupportedOperationException("Please call this method of the individual mutations");
   }
 
+  /**
+   * @return the highest priority of all the mutations
+   */
   @Override
-  public byte[] getAttribute(String name) {
+  public int getPriority() {
     if (action instanceof Mutation) {
-      return ((Mutation) action).getAttribute(name);
+      return ((Mutation) action).getPriority();
     }
-    throw new UnsupportedOperationException();
+    return ((RowMutations) action).getMaxPriority();
   }
 
   @Override
-  public OperationWithAttributes setAttribute(String name, byte[] value) {
-    if (action instanceof Mutation) {
-      return ((Mutation) action).setAttribute(name, value);
-    }
-    throw new UnsupportedOperationException();
+  public CheckAndMutate setPriority(int priority) {
+    throw new UnsupportedOperationException("Please call this method of the individual mutations");
   }
 
   @Override
-  public int getPriority() {
-    if (action instanceof Mutation) {
-      return ((Mutation) action).getPriority();
+  public CheckAndMutate setAttribute(String name, byte[] value) {
+    return (CheckAndMutate) super.setAttribute(name, value);
+  }
+
+  @Override
+  public String getId() {
```

Comment:
I want to keep CheckAndMutate extend Mutation because I want to treat CheckAndMutate as same as other mutations (Put/Delete/Increment/Append) to perform CheckAndMutate and other mutations atomically.

For example, after this change, we will be able to perform Put and Increment and CheckAndMutate atomically in a single row (or region).

However, I'm thinking that we need to rearrange/redesign the mutation structure in the future.